File: android/aurakit/src/main/kotlin/ai/aura/kit/DeepLink.kt
package ai.aura.kit

import java.net.URI
import java.net.URLDecoder

/**
 * Разбор ссылок схемы aura:// (URL-схема приложения) — тот же набор маршрутов,
 * что на iOS (DeepLink.swift) и в docs/MOBILE_FEATURES.md:
 *
 *   aura://chats/{id}                        — открыть чат
 *   aura://voice                             — голосовой запрос
 *   aura://ask?text=...                      — сразу спросить Ауру
 *   aura://tasks                             — список задач
 *   aura://settings                          — настройки
 *   aura://oauth?provider=&code=&state=...   — возврат из Google OAuth
 *
 * Неизвестные пути → null (вызывающий код их игнорирует, без краша).
 */
sealed class DeepLink {
    data class Chats(val id: Long?) : DeepLink()
    data object Voice : DeepLink()
    data class Ask(val text: String) : DeepLink()
    data object Tasks : DeepLink()
    data object Notifications : DeepLink()
    data object Settings : DeepLink()
    data class OAuth(val provider: String, val code: String, val state: String) : DeepLink()

    companion object {
        const val SCHEME = "aura"

        fun parse(url: String): DeepLink? {
            val uri = try {
                URI(url)
            } catch (_: Exception) {
                return null
            }
            if (!SCHEME.equals(uri.scheme, ignoreCase = true)) return null
            val host = uri.host?.lowercase() ?: return null
            val query = parseQuery(uri.rawQuery)

            return when (host) {
                "chats" -> {
                    val id = uri.path?.trimStart('/')?.substringBefore('/')?.toLongOrNull()
                    Chats(id)
                }
                "voice" -> Voice
                "ask" -> Ask(query["text"] ?: "")
                "tasks" -> Tasks
                "notifications" -> Notifications
                "settings" -> Settings
                "oauth" -> {
                    val provider = query["provider"] ?: return null
                    val code = query["code"] ?: return null
                    val state = query["state"] ?: return null
                    OAuth(provider, code, state)
                }
                else -> null
            }
        }

        private fun parseQuery(rawQuery: String?): Map<String, String> {
            if (rawQuery.isNullOrEmpty()) return emptyMap()
            return rawQuery.split('&').mapNotNull { pair ->
                val parts = pair.split('=', limit = 2)
                val key = parts.getOrNull(0)?.takeIf { it.isNotEmpty() } ?: return@mapNotNull null
                val value = parts.getOrNull(1) ?: ""
                key to URLDecoder.decode(value, Charsets.UTF_8)
            }.toMap()
        }
    }
}

File: ios/AuraKit/Sources/AuraKit/DeepLink.swift
// DeepLink.swift — разбор ссылок схемы aura:// (URL-схема приложения).
//
//   aura://chats/{id}                        — открыть чат
//   aura://voice                             — голосовой запрос (Back Tap → Команды)
//   aura://ask?text=...                      — сразу спросить Ауру
//   aura://tasks                             — список задач
//   aura://settings                          — настройки
//   aura://oauth?provider=&code=&state=...   — возврат из Google OAuth
//
// В продакшене схема дополняется Universal Links (https) — тот же роутинг.
import Foundation

public enum DeepLink: Equatable, Sendable {
    case chats(id: Int64?)
    case voice
    case ask(text: String)
    case tasks
    case settings
    case oauth(provider: String, code: String, state: String)

    public static let scheme = "aura"

    public init?(url: URL) {
        guard url.scheme?.lowercased() == DeepLink.scheme else { return nil }
        // URLComponents: host + компоненты пути и query.
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return nil
        }
        let host = components.host?.lowercased() ?? ""
        let queryItems = components.queryItems ?? []
        func query(_ name: String) -> String? {
            queryItems.first(where: { $0.name == name })?.value
        }

        switch host {
        case "chats":
            let id = components.path.split(separator: "/").first.flatMap { Int64($0) }
            self = .chats(id: id)
        case "voice":
            self = .voice
        case "ask":
            self = .ask(text: query("text") ?? "")
        case "tasks":
            self = .tasks
        case "settings":
            self = .settings
        case "oauth":
            guard let provider = query("provider"),
                  let code = query("code"),
                  let state = query("state") else { return nil }
            self = .oauth(provider: provider, code: code, state: state)
        default:
            return nil
        }
    }
}
